File: aoc2025/src/day_1.rs
use std::str::FromStr;

use aoclib::{LineParser, Runner};

#[derive(Default)]
pub struct Day1 {
    turns: Vec<Turn>,
}

impl Day1 {
    pub fn new() -> Self {
        Self::default()
    }
}

impl Runner for Day1 {
    fn name(&self) -> (usize, usize) {
        (2025, 1)
    }

    fn parse(&mut self) {
        let lines = aoclib::read_lines("aoc2025/input/day_1.txt");
        self.turns = lines.parse_lines();
    }

    fn part_one(&mut self) -> Vec<String> {
        let mut setting = 50;
        let mut counter = 0;
        for turn in &self.turns {
            match turn {
                Turn::Left(amount) => setting = (setting - amount).rem_euclid(100),
                Turn::Right(amount) => setting = (setting + amount) % 100,
            }
            if setting == 0 {
                counter += 1;
            }
        }
        aoclib::output(counter)
    }

    fn part_two(&mut self) -> Vec<String> {
        let mut setting = 50isize;
        let mut counter = 0;
        for turn in &self.turns {
            let (dir, mut amount) = match turn {
                Turn::Left(amnt) => (-1, *amnt),
                Turn::Right(amnt) => (1, *amnt),
            };

            while amount > 0 {
                setting = (setting + dir).rem_euclid(100);
                if setting == 0 {
                    counter += 1;
                }
                amount -= 1;
            }
        }

        aoclib::output(counter)
    }
}

#[derive(Debug)]
enum Turn {
    Left(isize),
    Right(isize),
}

impl FromStr for Turn {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let is_left = &s[0..1] == "L";
        let amount = s[1..].parse::<isize>().unwrap();
        Ok(if is_left {
            Turn::Left(amount)
        } else {
            Turn::Right(amount)
        })
    }
}

File: gopher/internal/day1/day1.go
package day1

import (
	"log"
	"strconv"

	"github.com/iiharsha/aoc/gopher/internal/aoclib"
)

type Day1 struct {
	turns []*turn
}

func (d *Day1) Name() (int, int) {
	return 2025, 1
}

func (d *Day1) Parse() {
	year, day := d.Name()
	lines, err := aoclib.ReadDayLines(year, day)
	if err != nil {
		log.Fatalf("err: %s -> %d", err, year)
	}

	d.turns = make([]*turn, 0, len(lines))
	for _, line := range lines {
		d.turns = append(d.turns, parseTurn(line))
	}
}

const startPos = 50

func (d *Day1) PartOne() []string {
	setting := startPos
	counter := 0
	for _, t := range d.turns {
		if t.dir == left {
			setting = remEuclid(setting-t.amount, 100)
		} else {
			setting = (setting + t.amount) % 100
		}

		if setting == 0 {
			counter++
		}
	}
	return aoclib.Output(counter)
}

func (d *Day1) PartTwo() []string {
	setting := startPos
	counter := 0
	for _, t := range d.turns {
		var dir = 1
		if t.dir == left {
			dir = -1
		}

		for t.amount > 0 {
			setting = remEuclid(setting+dir, 100)
			if setting == 0 {
				counter++
			}

			t.amount--
		}
	}

	return aoclib.Output(counter)
}

func remEuclid(a, b int) int {
	r := a % b
	if r < 0 {
		r += b
	}
	return r
}

type direction int

const (
	left direction = iota
	right
)

type turn struct {
	dir    direction
	amount int
}

func parseTurn(s string) *turn {
	amount, err := strconv.Atoi(s[1:])
	if err != nil {
		panic("error: converting input string number to integer")
	}

	if s[0] == 'L' {
		return &turn{dir: left, amount: amount}
	}
	return &turn{dir: right, amount: amount}
}
